### PR TITLE
Porting tern analysis of JS code to Node.

### PR DESCRIPTION
**Pull request by swmitra**
This PR is created to port our js code analysis logic on node for better scalability . This will improve the overall performance and leverage addition of new features like - 
1) In place rename of variable with scoping [ Separate PR will enable this soon ]
2) Find reference with scoping [ Separate PR will enable this soon ]
3) Call Hierarchy  [ Separate PR will enable this soon ]
## TODO
- [X] Fetching of doc content from node domain instead of posting request back to main thread
- [x] Unit test update

### CODE REVIEW

**Comment by swmitra** [comment]
Tagging @nethip @abose @zaggino @ryanstewart @petetnt

File: .gitmodules
Comment (abose) [inline .gitmodules]: two tern submodules. Can this be unified?

**Comment by swmitra** [comment]
@abose WIP! Removed the old reference now.

**Comment by abose** [comment]
Would the path change affect the extensions?

**Comment by petetnt** [comment]
@swmitra Have you done any initial performance comparisons against the previous implementation :bar_chart:?

I can try and review this ASAP, been a bit busy as of late

File: .gitmodules
Comment (ficristo) [inline .gitmodules]: Ooc, would not be better to create a `package.json` in `src/extensions/default/JavaScriptCodeHints/node`?

**Comment by ficristo** [comment]
@swmitra Do you think it would be possible / easier to add the reading of a `.tern-config` file (even in a future PR)?

**Comment by swmitra** [comment]
Sorry for late reply.

@petetnt Yes I tried to benchmark the performance to have the gain quantified (Will post a detailed result once the implementation is complete). 
Any help on performance data with and without this PR is welcome!
For small and medium JS files , there is marginal gain which may not be even perceptible. The benefits of porting this analysis to tern node has 2 main agendas - 
* File content extraction can be done from node domain itself. I am currently working on this. This will allow us to perform the complete analysis and preparation for the analysis outside main thread as we don't have to post back messages to the main thread asking for file content. 
* We do have a bottleneck in terms of performance when we are asking for hints in htmlmixed content and the html file is large enough ( may be 10k+ lines of code ). The typing response gets severely affected as the main thread gets hogged to extract JS text from the html file. I am trying to use a CM runmode on node to offload this to node domain. This will improve the typing performance of the main editor significantly. 

@ficristo Yes , it's a good idea to have a package.json and resolve the dependencies as node modules. Also we can introduce a tern config file. I will take both this activities outside this PR and do the changes under a separate PR soon after this.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: updateFilesCache

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: Any particular reason why the exports don't match the actual method names / should the methods themselves just be public (exporting private methods...)?

Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: Not any specific reason as such to have different names. Will do the change.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: @petetnt Sorry for late reply. Was busy with some the type details stuff :smiley: . I will add required JSDocs tonight.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: :+1:

File: .gitmodules
Comment (petetnt) [inline .gitmodules]: FYI `.gitmodules` now conflict with #11994 (and the `url` is wrong, should be `https://github.com/ternjs/tern.git`, but that will be fixed by fixing the conflict too)

**Comment by yacut** [comment]
Would be possible to enable nodejs plugin with this feature?

**Comment by ficristo** [comment]
@swmitra What is the status of this?

**Comment by swmitra** [comment]
I am really sorry @ficristo for late reply. 
This is not yet ready for delivery. Currently I am help with a lot of work. I will try to look at this next week.

**Comment by ficristo** [comment]
@swmitra could you update this to latest master?

File: .gitmodules
Comment (zaggino) [inline .gitmodules]: @swmitra 
we don't need submodules anymore, `npm install` will install dependencies defined in `src/extensions/default/JavaScriptCodeHints/package.json`, please remove

**Review by zaggino — CHANGES_REQUESTED** [review]

File: src/extensions/default/JavaScriptCodeHints/node/package.json
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/node/package.json]: @swmitra 
this script fails when trying to run `npm install`
I don't think it's needed if tern doesn't run in webworker anymore

**Comment by swmitra** [comment]
@zaggino Thanks a lot for the quick review.  I will update the PR with the requested changes.  As this PR was raised before the dependency handling change in Brackets, gitmodule change is still visible in change list.  I will make that even with master.

**Comment by zaggino** [comment]
This should be absolute must for 1.9 @swmitra , I believe that it will fix issues some people have with Brackets crashing on them when viewing certain files. cc @ficristo

**Comment by swmitra** [comment]
Sure @zaggino.  I will be on it.

**Comment by swmitra** [comment]
@zaggino I have removed the redundant module definitions and the post install script as well. Please have a look at the PR. Meanwhile I will do some unit testing around inference data clearance on project change ( that's the only bug remaining now).

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: please remove

File: .gitmodules
Comment (zaggino) [inline .gitmodules]: extra space here?

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: please remove

File: src/extensions/default/JavaScriptCodeHints/node/package.json
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/node/package.json]: latest acorn is `4.0.11`, can we upgrade this to it?

**Review by zaggino — CHANGES_REQUESTED** [review]
hi @swmitra 
been using this for a day or so and it seems to work fine without any crashes or errors, can you please clean-up this so it'll be ready to merge? also it'd be worth to see what breaking changes are in acorn 4.x and if we can upgrade to it now, rather than later

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (zaggino) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: this should be an object `{}` and used like `_dirtyFilesCache[fileName]` instead of searching an array

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (ficristo) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: `completions = data.completions.map(...` ?

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (ficristo) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: can this be something like `... .types.map(inferTypeToString).join(", ")`?

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (ficristo) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: Can you use 
```js
Object.keys(pros).map(function (key)  {
 return key + ": " + inferTypeToString(props[key]);
}).join(", ")
```

File: src/extensions/default/JavaScriptCodeHints/node/package.json
Comment (ficristo) [inline src/extensions/default/JavaScriptCodeHints/node/package.json]: I didn't see `acorn` used on node side, is it?
Also in the latest releases there were some problems, I'd upgrade it in a separate PR when things are settled down on `acorn` side.

**Review by ficristo — DISMISSED** [review]
When I tested this branch a little the autocomplete didn't seem really robust enough. Sometime I had some empty items in the autocomplete list.
Also all manual handling in `TernNodeDomain.js` seemed a bit scary but I suppose it is necessary to convert messages from a format to another.

File: src/extensions/default/JavaScriptCodeHints/node/package.json
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/package.json]: Acorn is used inside Tern by the Infer module. I agree with @ficristo, we will upgrade Acorn after a stable release.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: Great catch! Infact this was a bug.

**Comment by ficristo** [comment]
While I believe this is a must for us, I think this will break js-code-hints the in-browser approach.
Maybe there is something we could do to be less problematic for them?
/cc @humphd

**Comment by humphd** [comment]
@ficristo thanks for alerting me to this (and if you see other things that you think might break us, I'd appreciate being cc'ed again).  If we have to, we can just fork the existing extension and use that, while you move forward with a node-based approach.

**Comment by zaggino** [comment]
@swmitra I believe last comment here: https://github.com/adobe/brackets/issues/12617#issuecomment-281417766 proves that this is an absolute must to merge for 1.9

**Comment by swmitra** [comment]
@zaggino @ficristo @petetnt Can you guys please have a look at the code now? If there are any breaking changes or adverse effect on brackets, we can move this to 1.10 milestone and unblock the 1.9 release. 
We are already in the process of 1.9 release activities and having a release of Brackets is an absolute must as it's over four months we did the last release. Our goal is to have a release every 2 months from now on.

**Comment by swmitra** [comment]
@zaggino @petetnt Can you guys please have a look at this? 
This is a pending feature for 1.10. I have resolved all the outstanding issues/bugs and JS hinting, jump to def all are working fine.

**Comment by petetnt** [comment]
@swmitra running the branch right now at work, I can give some feedback later.

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: This is just a check to tell the type. If we can't determine the type using tern we send the original response back.

**Comment by petetnt** [comment]
Been using this branch for couple of days without issues **except** CMD+O (quick open) won't open files that are already open in the working set, which is a regression from the current behaviour. Haven't had the time to dive in whats up with that, but FYI 👍

**Review by zaggino — APPROVED** [review]

File: src/extensions/default/JavaScriptCodeHints/node/package.json
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/package.json]: Latest tern is now `0.21.0`. If we can update to that, great. If it has breaking changes, lets keep this one and update later.

**Review by petetnt — APPROVED** [review]
LGTM with a notice

**Comment by swmitra** [comment]
Pushing this to next release. We will merge this the moment we release 1.10 and do the first pre-release with this PR. It's breaking a few unit tests and have to resolve them before we merge.

**Comment by MarcelGerber** [comment]
@swmitra Based on your last comment, this is now ready to be merged.

**Comment by swmitra** [comment]
Yes @MarcelGerber 😄 . In case we need a build refresh for the reported dll issue,  we would be in trouble in case this PR is merged. I would wait for a couple of days to discuss any possibility of build refresh and then merge this PR for the first pre release.

**Comment by swmitra** [comment]
@MarcelGerber This PR can't be merged right now 😞.  There are 2 major issues I found today while testing - 
## TOFIX
- [ ] Require module hints have stopped working 
- [ ] If node restarts for any failure, re initialization of Tern should happen again 

While I am working on these 2 issues, if you can spot any problem or a probable cause for this, please let me know. These should be fixed in a day or two.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: Now that Brackets handles arbitary encodings, should the encoding be passed to this function too?

Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: Precisely I am in a catch 22 situation 😢 . Encoding scheme is only available after detection at she'll level, in order to use that, I have to read all the files by hopping on the Brackets main thread, which negates any gain in terms of performance compared to using nodes io. I was relying on Brackets only to fetch dirty files content, but If I start relying on Brackets we will end up with performance issues in case of flat project structure. Prime pump will clog the main thread which in turn will make the entire editor non responsive.

**Comment by swmitra** [comment]
Acorn and Tern have been upgraded to latest version. Haven't noticed any problem in real use or test suite except the ones mentioned before.

File: src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/ExtractFileContent.js]: @petetnt I have fixed most of the issues. Relying on Brackets main thread for content until we figure out a way to pass encoding info.

**Comment by swmitra** [comment]
@MarcelGerber @petetnt Just the final bit remaining. Need to re-initialize Tern when node gets restarted after a crash.

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (petetnt) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: Do we need this?

**Review by petetnt — COMMENTED** [review]
One more question

File: src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js
Comment (swmitra) [inline src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js]: Yes Pete. Without these, commonjs dependency analysis and js doc parsing for parameter type inference doesn't work.

**Comment by petetnt** [comment]
LGTM with a follow up PR on the restart-after-crash scenario

**Pull request merged**
